### ISSUE TITLE
Add raylib backend ?

### ISSUE THREAD

**Issue by Lecrapouille** [ocornut/imgui#6933]
Hi! This is just a suggestion to add an official backend for raylib. @JeffM2501 what do you suggest?

One of the interesting point of raylib is that it can be compiled for several OS and especially with Emscripten which makes the backend architecture agnostic (for example not switching between either opengl3 or wgpu).

- https://github.com/raylib-extras/rlImGui
- https://github.com/oskaritimperi/imgui-impl-raylib
- https://www.raylib.com/

**Comment by ocornut** [comment]
There’s already a raylib backend and you just linked to it. I see no value in me taking over maintainance of an existing backend.

**Comment by Lecrapouille** [comment]
@ocornut No worries it was just a suggestion because the first link seems not respecting your API convention (ie ImGui_ImplRaylib_Init =>rlImGuiSetup) the second is respecting but seems outdated. I was just thinking about basic one-to-one function wrapper.

**Comment by JeffM2501** [comment]
I am the author of rlImGui, it is well maintained ;)
We have C++, C#, and C compatibility.
rlImGui is used by the raylib community extensively.

**Comment by ocornut** [comment]
I also find it a little unfortunate that backends are created not following our conventions, but I do understand from the pov of an ecosystem manager they are aiming to steer the coding conventions toward theirs. In the end its only a few functions call and shouldn’t matter too much.

**Comment by ocornut** [comment]
The naming rlImGuiBegin for NewFrame() when we have an ImGui::Begin() function that does a different thing is indeed quite awkward.

Maybe for the sake of documentation they should be commented to refer to the names dear imgui backends uses.

**Comment by JeffM2501** [comment]
The goal of rlImGui is to fit more with the raylib API for sure. Raylib is first and foremost designed as a learning library to help people who are just starting with gamedev, so the goal is to keep the API simple. Raylib uses a lot of Begin/End functions to start and stop modes, and this fit right in. It's main goal is to allow ImGui integration with 4 simple function calls.

I will for sure comment up the functions.

I will also make an Imgui compatible interface for those that are more used to that style.

@Lecrapouille you should have just brought your concerns up to me to begin with :)
I do not think that Omar or ImGui should talk on the maintenance of the raylib backend, it has some very specific useability goals, and I need Omar to work on issues for my real job ;)

**Comment by JeffM2501** [comment]
rlImGui has been updated with a low level API header that conforms to the convention set forth by other backends (I based it on SDL2). I also added an example using that API and commented up the high level API to mention what low level API functions it calls.

**Comment by Lecrapouille** [comment]
Sorry I was busy yesterday Jeff I will look at your example tonight :) I hope you did not break your current API for my ticket.

And sorry guys I did not want to create conflicts, just trying to make a more unified AP. A basic one-to-one function wrapper: `rlImGuiBegin() { NewFrame(); }` is enough. I did not know in which repo to discuss better unification:
- I know, as Dear imgui beginner, the number of begin/end in the correct order is sometimes confusing (while assert help a lot) and I understand the desire to have a higher level of abstraction.
- On other point, I understand the extra work to Ocornut to maintain such wrapper.
